File: internal/infra/di/update_payment_status.go
package di

import (
	driven_dynamodb "github.com/fiap/ms-payment/internal/adapters/driven/dynamodb"
	driven_mercadopago "github.com/fiap/ms-payment/internal/adapters/driven/mercadopago"
	driven_sns "github.com/fiap/ms-payment/internal/adapters/driven/sns"
	"github.com/fiap/ms-payment/internal/core/application/usecases"
	infra_aws "github.com/fiap/ms-payment/internal/infra/aws"
)

type UpdatePaymentStatusContainer struct {
	UpdatePaymentStatusUseCase *usecases.UpdatePaymentStatusUseCase
}

func NewUpdatePaymentStatusContainer(region infra_aws.AwsRegion, stage infra_aws.AwsStage) *UpdatePaymentStatusContainer {
	dynamoClient := infra_aws.NewDynamoDBClient(region, stage)
	smClient := infra_aws.NewSMClient(region, stage)
	snsClient := infra_aws.NewSNSClient(region, stage)

	paymentRepo := driven_dynamodb.NewPaymentRepository(*dynamoClient)
	generateTokenMP := driven_mercadopago.NewMercadoPagoGenerateToken(smClient)
	getPaymentMP := driven_mercadopago.NewMercadoPagoGetPayment()
	notificationService := driven_sns.NewPaymentSNS(snsClient)

	return &UpdatePaymentStatusContainer{
		UpdatePaymentStatusUseCase: usecases.NewUpdatePaymentStatusWebhookUseCase(
			paymentRepo,
			generateTokenMP,
			getPaymentMP,
			notificationService,
		),
	}
}

File: src/infra/di/update_payment_status.ts
import { MercadoPagoGenerateToken } from "@driven_mercadopago/generate_token"
import { UpdatePaymentStatusUseCase } from "@usecases/update_payment_status"
import { PaymentRepository } from "@driven_dynamodb/payment_repository"
import { MercadoPagoGetPayment } from "@driven_mercadopago/get_payment"
import { DynamoDBClientWrapper } from "@aws/dynamodb_client"
import { SNSClientWrapper } from "@aws/sns_client"
import { AwsRegion, AwsStage } from "@aws/utils"
import { SMClientWrapper } from "@aws/sm_client"
import { PaymentSNS } from "@driven_sns/payment"

export class UpdatePaymentStatusContainerFactory {
	usecase: UpdatePaymentStatusUseCase

	constructor(region: AwsRegion, stage: AwsStage) {
		const dynamoClient = new DynamoDBClientWrapper(region, stage)
		const smClient = new SMClientWrapper(region, stage)
		const snsClient = new SNSClientWrapper(region, stage)

		const paymentRepo = new PaymentRepository(dynamoClient)
		const generateTokenMP = new MercadoPagoGenerateToken(smClient)
		const getPaymentMP = new MercadoPagoGetPayment()
		const notificationService = new PaymentSNS(snsClient)

		this.usecase = new UpdatePaymentStatusUseCase(
			paymentRepo,
			generateTokenMP,
			getPaymentMP,
			notificationService,
		)
	}
}
